backends/ebpf/tests/ptf/common.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2022-present Orange
+# Copyright 2022-present Open Networking Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import logging
+import json
+import shlex
+import subprocess
+import ptf
+import ptf.testutils as testutils
+
+from ptf.base_tests import BaseTest
+
+logger = logging.getLogger('eBPFTest')
+if not len(logger.handlers):
+    logger.addHandler(logging.StreamHandler())
+
+TEST_PIPELINE_ID = 1
+TEST_PIPELINE_MOUNT_PATH = "/sys/fs/bpf/pipeline{}".format(TEST_PIPELINE_ID)
+PIPELINE_MAPS_MOUNT_PATH = "{}/maps".format(TEST_PIPELINE_MOUNT_PATH)
+
+def xdp2tc_head_not_supported(cls):
+    if cls.xdp2tc_mode(cls) == 'head':
+        cls.skip = True
+        cls.skip_reason = "not supported for xdp2tc=head"
+    return cls
+
+
+class P4EbpfTest(BaseTest):
+    """
+    Generates BPF bytecode from a P4 program and runs a PTF test.
+    """
+
+    skip = False
+    skip_reason = ''
+    switch_ns = 'test'
+    p4_file_path = ""
+
+    def setUp(self):
+        super(P4EbpfTest, self).setUp()
+
+        if self.skip:
+            self.skipTest(self.skip_reason)
+
+        if not os.path.exists(self.p4_file_path):
+            self.fail("P4 program not found, no such file.")
+
+        if not os.path.exists("ptf_out"):
+            os.makedirs("ptf_out")
+
+        head, tail = os.path.split(self.p4_file_path)
+        filename = tail.split(".")[0]
+        self.test_prog_image = os.path.join("ptf_out", filename + ".o")
+
+        p4args = "--Wdisable=unused"
+        if self.is_trace_logs_enabled():
+            p4args += " --trace"
+
+        logger.info("P4ARGS=" + p4args)
+        self.exec_cmd("make -f ../runtime/kernel.mk BPFOBJ={output} P4FILE={p4file} "
+                      "ARGS=\"{cargs}\" P4C=p4c-ebpf P4ARGS=\"{p4args}\" psa".format(
+                            output=self.test_prog_image,
+                            p4file=self.p4_file_path,
+                            cargs="-DPSA_PORT_RECIRCULATE=2",
```

Comment:
since this is defined here perhaps you can use the macro in the P4 code.

Reply:
I've check that, but if we define the macro in the P4 code it's not propagated down to the C code. We must pass it to `clang`. Also, hardcoding it in the P4 code would require modifying P4 programs each time the recirculate port number is changed (it might become dynamically allocated in future). 

Please also note that the packet recirculation support should be rather treated as an "experimental" feature. We'll create an issue to track this.